### PR TITLE
new: Add support for LKE Control Plane ACLs

### PR DESCRIPTION
**Pull request by lgarber-akamai**
## 📝 Description

This pull request adds support for configuring and viewing the ACL configuration for an LKE cluster's control plane.

**NOTE: This PR does NOT include the PUT LKE cluster changes because the acl configuration is not returned in the GET LKE cluster response.**

#### New Endpoint Methods

* `control_plane_acl` - GET `/lke/clusters/{cluster_id}/control_plane_acl`
* `control_plane_acl_update(...)` - PUT `/lke/clusters/{cluster_id}/control_plane_acl`
* `control_plane_acl_delete()` - POST `/lke/clusters/{cluster_id}/control_plane_acl`

#### Updated Endpoint Methods

* `LKEGroup.cluster_create(...)` - Add control_plane field to method arguments

#### Misc Changes

* Added data classes for LKE cluster control plane and all substructures
* Added logic to JSONObject to remove `Optional` values from the generated dict if None
    * Added a new `always_include` class var used to designate optional values that should always be included in the generated Dict
* Updated test fixture framework to support underscores in path

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally and run `make install`.

### Unit Testing

```
make testunit
```

### Integration Testing

```
make TEST_COMMAND=models/lke/test_lke.py testint
```

### Manual Testing

In a Python SDK sandbox environment (e.g. dx-devenv), run the following:

```python
import os

from linode_api4 import (
    LinodeClient,
    LKEClusterControlPlaneOptions,
    LKEClusterControlPlaneACLOptions,
    LKEClusterControlPlaneACLAddressesOptions,
)

client = LinodeClient(token=os.getenv("LINODE_TOKEN"))

cluster = client.lke.cluster_create(
    "us-mia",
    "test-cluster",
    client.lke.node_pool("g6-standard-1", 1),
    "1.29",
    control_plane=LKEClusterControlPlaneOptions(
        acl=LKEClusterControlPlaneACLOptions(
            enabled=True,
            addresses=LKEClusterControlPlaneACLAddressesOptions(
                ipv4=["10.0.0.1/32"], ipv6=["1234::5678"]
            ),
        )
    ),
)

print("Original ACL:", cluster.control_plane_acl)

cluster.control_plane_acl_update(
    LKEClusterControlPlaneACLOptions(
        addresses=LKEClusterControlPlaneACLAddressesOptions(ipv4=["10.0.0.2/32"])
    )
)

print("Updated ACL:", cluster.control_plane_acl)

cluster.control_plane_acl_delete()

print("Deleted ACL:", cluster.control_plane_acl)
```

2. Ensure the script runs successfully and the output matches the following:

```
Original ACL: LKEClusterControlPlaneACL(enabled=True, addresses=LKEClusterControlPlaneACLAddresses(ipv4=['10.0.0.1/32'], ipv6=['1234::5678/128']))
Updated ACL: LKEClusterControlPlaneACL(enabled=True, addresses=LKEClusterControlPlaneACLAddresses(ipv4=['10.0.0.2/32'], ipv6=None))
Deleted ACL: LKEClusterControlPlaneACL(enabled=False, addresses=None)
```

### CODE REVIEW

File: linode_api4/objects/serializable.py [hunk 1]
```diff
@@ -117,7 +156,45 @@ def _serialize(self) -> Dict[str, Any]:
...
+            if isinstance(v, List):
+                v = [attempt_serialize(j) for j in v]
+            elif isinstance(v, Dict):
+                v = {k: attempt_serialize(j) for k, j in v.items()}
+            else:
+                v = attempt_serialize(v)
+
+            result[k] = v
+
+        return result
```
Comment (lgarber-akamai) [inline linode_api4/objects/serializable.py]: Unfortunately we have to reinvent the wheel here instead of using `asdict(...)` because we need to differentiate between Optional and Nullable values for certain edge cases. 

Ideally I'd want to break this out into a separate PR but excluding optional fields from generated request bodies is a requirement for `LKEClusterControlPlaneACL`.

File: test/unit/fixtures.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
 import json
 import os
+import re
 import sys
 
 FIXTURES_DIR = sys.path[0] + "/test/fixtures"
 
+# This regex is useful for finding individual underscore characters,
+# which is necessary to allow us to use underscores in URL paths.
+PATH_REPLACEMENT_REGEX = re.compile(r"(?<!_)_(?!_)")
```
Comment (lgarber-akamai) [inline test/unit/fixtures.py:10]: This is necessary for `lke/clusters/18881/control_plane_acl` fixture since the `_`s were being replaced with `/`s.

File: linode_api4/objects/lke.py [hunk 3]
```diff
@@ -335,3 +407,40 @@ def service_token_delete(self):
         self._client.delete(
             "{}/servicetoken".format(LKECluster.api_endpoint), model=self
         )
+
+    def control_plane_acl_update(
```
Comment (lgarber-akamai) [inline linode_api4/objects/lke.py:411]: This pattern is not consistent with the rest of the codebase, but unfortunately we cannot implement it using the traditional properties because `ACL` is not returned from the LKE cluster GET endpoint.

File: linode_api4/objects/vpc.py [hunk 4]
```diff
@@ -100,7 +100,7 @@ def subnet_create(
         return d
 
     @property
-    def ips(self, *filters) -> PaginatedList:
+    def ips(self) -> PaginatedList:
```
Comment (lgarber-akamai) [inline linode_api4/objects/vpc.py:103]: This isn't directly related to this change but it was causing the linter to fail. This shouldn't be a breaking change since arguments could already not be passed into ips 👍

File: linode_api4/objects/serializable.py [hunk 5]
```diff
@@ -54,28 +57,57 @@ class JSONObject(metaclass=JSONFilterableMetaclass):
         )
     """
 
+    always_include: ClassVar[Set[str]] = {}
+    """
+    A set of keys corresponding to fields that should always be
+    included in the generated output regardless of whether their values
+    are None.
+    """
```
Comment (lgarber-akamai) [inline linode_api4/objects/serializable.py:65]: I'm wondering if it'd make more sense to do something more similar to `omitempty` like in Go.

Comment (lgarber-akamai) [inline linode_api4/objects/serializable.py:65]: I was considering implementing something more similar to `omitempty` in Go, but in this case I think there are far fewer instances where we would want to include null values rather than excluding them.

File: linode_api4/objects/serializable.py [hunk 6]
```diff
@@ -117,7 +151,55 @@ def _serialize(self) -> Dict[str, Any]:
...
+            if isinstance(v, List):
+                v = [attempt_serialize(j) for j in v]
+            elif isinstance(v, Dict):
+                v = {k: attempt_serialize(j) for k, j in v.items()}
+            else:
+                v = attempt_serialize(v)
+
+            result[k] = v
+
+        return result
```
Comment (lgarber-akamai) [inline linode_api4/objects/serializable.py:202]: Unfortunately we have to reinvent the wheel here because we don't have access to type hints or `always_include` for nested JSONObjects in the asdict `dict_factory` method.

Ideally I'd want to break this out into a separate PR but excluding optional fields from generated request bodies is a requirement for `LKEClusterControlPlaneACL`.

File: linode_api4/objects/lke.py [hunk 7]
```diff
@@ -26,6 +29,61 @@ class KubeVersion(Base):
...
+class LKEClusterControlPlaneACL(JSONObject):
+    """
+    LKEClusterControlPlaneACL describes the ACL configuration of an LKE cluster's
+    control plane.
+    """
+
+    enabled: bool = False
+    addresses: LKEClusterControlPlaneACLAddressesOptions = None
+
+
```
Comment (lgarber-akamai) [inline linode_api4/objects/lke.py:86]: This is more of a linodego pattern but I think it's necessary given we're starting to see more diverging API request & response structures

**Review by yec-akamai — APPROVED** [review]
Tested and everything passed locally, nice work!

File: linode_api4/objects/lke.py (hunk 7 above)
Comment (yec-akamai) [reply linode_api4/objects/lke.py]: I'm wondering if we should separate them in ansible and tf as well. One of the use case that I can think of is in terraform framework, nested block can't be both optional and computed. It's a bit annoying when doing the migration. We did the customized object type, but Hashicorp once suggested that we could have two separated block attributes, one is optional and one is computed to implement that.

Comment (lgarber-akamai) [reply linode_api4/objects/lke.py]: > One of the use case that I can think of is in terraform framework, nested block can't be both optional and computed. It's a bit annoying when doing the migration. We did the customized object type, but Hashicorp once suggested that we could have two separated block attributes, one is optional and one is computed to implement that.

That's a good call out! I think in the long term we'll need to split apart user-defined and computed structures anyway because of the limitations in TF framework you mentioned

**Review by ykim-1 — APPROVED** [review]
LGTM, tests pass locally

**Comment by ykim-1** [comment]
Also, thanks for fixing the lint error 👍
